src/fsharp/FSharp.Core/seq.fs

```diff
@@ -1577,57 +1577,56 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Sum")>]
         let inline sum (source: seq< (^a) >) : ^a = 
-            use e = source.GetEnumerator() 
             let mutable acc = LanguagePrimitives.GenericZero< (^a) >
-            while e.MoveNext() do
-                acc <- Checked.(+) acc e.Current
+            for item in source do
+                acc <- Checked.(+) acc item
```

Comment:
But "source" is not known to be an array, at least not known to the compiler - it is a sequence.  I think this applies for inlined code as well, even if at the callsite the thing is known to be an array (that's worth checking though).  

So the "for" loop over a sequence is effectively the same as the "while" loop.

At least, take a look at the perf and/or the IL code generated for callers and check if there's a substantive difference?

thanks